### PR TITLE
feat(balance): uncanny dodge can now ninja roll to dodge bullets

### PR DESCRIPTION
**Pull request by scarf005**
## Purpose of change

image

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/8e2ba940-f7d1-42ec-b3ab-8351c179e878

it's #4125 but with ninja moving

## Describe the solution

on top of #4125, it adds...

- `src/cata_algo.h`: a very [convoluted template mess to use `std::variant` sanely](https://en.cppreference.com/w/cpp/utility/variant/visit)!

- `src/character_functions.h`: describe uncanny dodge simulation result in `std::variant`!
- `src/character_functions.cpp`:
	- separate uncanny dodge simulation and execution into `uncanny_dodge_result` and `try_uncanny_dodge`!
	- creates smoke when dodging to block sight and reduce adjacent tiles to dodge when `thruster mode` is active

## Testing

1. edit bionics, give max power and uncanny dodge CBM
2. spawn turret.
3. set dodge to 0, check that you dodge the bullets around 50% of the time
4. set dodge to 10, check that you dodge the bullets 100% of the time
5. set dodge to 0, enable `Uncanny Dodge Thrusters Mode`, check that you dodge the bullets 100% of the time with leaving smoke behind (unless you're not surrounded by unsafe terrain)
6. set dodge to 10, enable `Uncanny Dodge Thrusters Mode`, check that you dodge the bullets 100% of the time without leaving smoke behind
7. set dodge to 0, enable `Uncanny Dodge Thrusters Mode`, stand inside 3x3 (or larger) smoke/fire, check that you fail to dodge via thrusters
8. set dodge to 0, enable `Uncanny Dodge Thrusters Mode`, enable debug HVAC, stand inside 3x3 (or larger) smoke/fire, check that you can dodge via thrusters

## Describe alternatives you've considered

- go with #4125, that's also perfectly fine solution.
- make thrusters mode separate CBM? (since currently it's a bit awkward as at dodge level 10 thruster mode does nothing)

## Additional Context

image

### CODE REVIEW

**Review by chaosvolt — CHANGES_REQUESTED** [review]
I'm warming up to this version of it, but only thing I can add feedback-wise is either:
1. The CBM descriptions and dodge message should give some hint that the smoke appearing on dodge is actually intentional and not some obscure bug players are going to complain about.
2. Or the smoke should not be implemented and instead rely on its high trigger cost as the main limiting factor.

EDIT: Also test fail due to something going on here:
```
Error: src/character_functions.cpp:646:9: error: control reaches end of non-void function [-Werror=return-type]
  646 |         }
      |         ^
```

**Comment by chaosvolt** [comment]
An additional thought follows after more discussion on the discord: if the player is trying to make a beeline from cover to cover, the randomized moves might end up becoming a giant annoyance by way of moving them every which way except where they're actually trying to go. This brings to mind fun situations where getting tagged once by a riot turret tends to end in you eating the entire cylinder because the beanbag round's stun effect causes you to bumblefuck about in the open constantly instead of stepping back into cover.

**Comment by scarf005** [comment]
how about we reintroduce dodge skill, but as a fallback measure:

- find safe adjacent tile
	- succeeds: roll an easy dodge check
		- succeeds: dodge the attack and stay in-place
		- fails: dodge the attack but move to that safe tile
	- all adjacent tiles are dangerous: roll a medium difficulty dodge check
		- succeeds: dodge the attack but stagger randomly into the dangerous adjacent tile
		- fails: do nothing and get hit
	- you are surround in walls in all direction
		- do nothing and get hit

**Comment by chaosvolt** [comment]
Hmm. Having dodge roll as a fallback may work but make it more complicated, if you want then. Definitely axe the ninja vanish smoke though given the potential issues with it though?

**Comment by scarf005** [comment]
or... toggleable option to  prefer approaching/move back from shooter?

**Comment by chaosvolt** [comment]
> or... toggleable option to prefer approaching/move back from shooter?

Seems like extra fluff and keypresses trying to guess the player's intentions as opposed to just not messing with the player's control over their character. :/

File: src/character_functions.h [hunk 1]
```diff
@@ -147,14 +148,31 @@ void store_in_container( Character &who, item &container, detached_ptr<item> &&p
...
  */
 bool try_uncanny_dodge( Character &who );
 
-/** Returns an unoccupied, safe adjacent point. */
-std::optional<tripoint> pick_safe_adjacent_tile( const Character &who );
+/**
+ * Returns an unoccupied, safe adjacent point.
+ *
+ * @warning this won't take account of whether you're immune to dangerous fields.
+ * TODO: add boolean flag to take account of player immunity?
```
Comment (Coolthulhu) [inline src/character_functions.h:173]: `bool` flags are considered bad practice, it's usually better to create two functions, in this case something like `_with_immunities` or something.

File: src/cata_algo.h [hunk 2]
```diff
@@ -159,6 +159,17 @@ auto group_by( const C &c, F && selector )
...
+/// @brief helper struct to pattern match `std::variant` variants
+///
+/// @remark poor person's https://doc.rust-lang.org/rust-by-example/flow_control/match.html
+///
+/// @see https://en.cppreference.com/w/cpp/utility/variant/visit
+/// @see https://polomack.eu/std-variant/
+template <class... Ts> struct match : Ts... {
+    using Ts::operator()...;
+};
+template <class... Ts> match( Ts... )->match<Ts...>;
```
Comment (Coolthulhu) [inline src/cata_algo.h:171]: Usage example?

File: src/character_functions.cpp [hunk 3]
```diff
@@ -592,34 +593,62 @@ bool try_wield_contents( Character &who, item &container, item *internal_item, b
...
-        return true;
-    } else {
-        if( is_u ) {
-            add_msg( _( "You try to dodge but there's no room!" ) );
-        } else if( seen ) {
-            add_msg( _( "%s tries to dodge but there's no room!" ), who.disp_name() );
+
+    const auto visitor = cata::match{
+        [&who, is_u, seen]( const tripoint & dest )
+        {
```
Comment (Coolthulhu) [inline src/character_functions.cpp:622]: No dodge checks of any sort?

File: src/cata_algo.h (hunk 2 above)
Comment (scarf005) [reply src/cata_algo.h:171]: used in:

https://github.com/cataclysmbnteam/Cataclysm-BN/pull/4139/files/514c2c5dd5abfca9cae1b0b1cc05bf207c449810#diff-53d614668bf7e514042903ad1eaaea53a9b7e018ac3c827c27ff6bb3b8616550R620-R648

this template is direct port from https://en.cppreference.com/w/cpp/utility/variant/visit#Example.

File: src/character_functions.cpp (hunk 3 above)
Comment (scarf005) [reply src/character_functions.cpp:645]: reintroduced in [`82559f1` (#4139)](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/4139/commits/82559f138ba7e458bf423ec9d30a4c9883e3ed4b)

File: src/character_functions.h (hunk 1 above)
Comment (scarf005) [reply src/character_functions.h]: fixed in [`deeae9a` (#4139)](https://github.com/cataclysmbnteam/Cataclysm-BN/pull/4139/commits/deeae9a2adcfdc4074588a54bccd633882144c11)

**Review by chaosvolt — COMMENTED** [review]
1. Compiled and load-tested.
2. Debugged myself a full powergen setup, power storage, and uncanny dodge CBM.
3. Spawned in a turret and turned the 
4. You no longer get a message or any other effect indicating a successful dodge, the turret just misses
5. In addition, it seems to be missing 100% when at 0 dodge skill, even when you're literally adjacent to the turret. This is presumably because having the CBM on itself gives you a dodge bonus to make it actually useful in melee, combined with how effective dodge is higher than dodge skill.

Thing is, I dunno exactly what changes to request because I dunno how much of the PR OP is still meant to be current intended behavior instead of olkd text that's no longer relevant.

If you intend for dodge skill to factor into the equation at a specific rate of "50% miss chance at zero dodge skill" then likely you'll want to call `get_skill_level( skill_dodge )` instead of `get_dodge()`.

**Comment by chaosvolt** [comment]
Wait. For some reason github desktop was several commits out of date, one sec while I retest

**Review by chaosvolt — CHANGES_REQUESTED** [review]
There we go, now it seems to work as the PR OP states, sorry about that.

1. Compiled and load-tested.
2. Installed debug fuel cell bionic, uncanny dodge, and 4 km-2 power storage CBMs.
3. Spawned in a turret and turned on the normal mode.
4. After the first successful dodge, the next shot claims I have no room and yet I still successfully dodged the shot anyway.
5. Having a thruster mode as a separate CBM also does not work right. It automatically powers down a turn or two after being enabled, regardless of whether you've turned the normal mode on or not.

EDIT: Remembered to add a screenshot this time, shows the thrusts failing to stay on, plus this time I got shot while standing in the open after it printed the failure message twice, instead of printing it once then dodging anyway:
image

My general thoughts:
1. I'd say axe the secondary bionic idea, it's gonna be a pain in the ass to implement because it needs to only be on when the base mode is on, plus ideally turning the thruster mode on would by necessity turn the base mode on at the same time for the sake of player convenience.
2. Instead I'd suggest either focusing on roll to dodge without displacing the player to a random tile, or perfect dodges that do displace you to a random tile and leave behind a fitting dangerous field. Per Coolthulhu's feedback I assume the former would be more desirable.
3. Obviously fix message oddities, but if displacement dodges get axed in favor of skill rolls then that would logically fix the "tells you you can't dodge anywhere but still dodges the shot anyway"
4. Optional idea: if we want a limiting factor on max bullet dodges that doesn't involve movement restrictions, we could instead have dodging check for number of dodges left as with melee dodges, and have successful bullet dodges reduce them. This would be default leave normal characters able to dodge only one shot per turn and thus make automatic fire more dangerous while synergizing well with martial arts that grant bonus dodges. Alternatively, we could make the CBM grant a bonus dodge or two by default while active if we want "can try to dodge more than one shot per turn" the default behavior.

File: src/character_functions.cpp [hunk 4]
```diff
@@ -595,63 +599,140 @@ bool try_wield_contents( Character &who, item &container, item *internal_item, b
...
+    if( who.get_power_level() < bio_uncanny_dodge->power_trigger ) {
+        return UncannyDodgeStatus::NoEnergy;
+    }
+
+    if( x_in_y( 10 + who.get_dodge(), 20 ) ) {
+        return UncannyDodgeStatus::DodgedWithSkill;
+    }
+
+    return UncannyDodgeStatus::NoSpace;
+}
```
Comment (chaosvolt) [inline src/character_functions.cpp:627]:
```suggestion
auto uncanny_dodge_result( const Character &who ) -> UncannyDodgeResult
{
    if( who.get_power_level() < bio_uncanny_dodge->power_trigger ) {
        return UncannyDodgeStatus::NoEnergy;
    }

    if( x_in_y( 10 + who.get_dodge(), 20 ) ) {
        return UncannyDodgeStatus::DodgedWithSkill;
    }

    return UncannyDodgeStatus::Fail;
}
```

File: src/character_functions.cpp [hunk 5]
```diff
@@ -595,63 +599,140 @@ bool try_wield_contents( Character &who, item &container, item *internal_item, b
...
+                add_msg( _( "Time seems to slow down and you're ejected to safe space!" ) );
+            } else if( seen ) {
+                add_msg( _( "%s dodges so fast as if there's rocket attached behind!" ), who.disp_name() );
+            }
+            auto &here = get_map();
+            here.add_field( who.pos(), fd_smoke, 1 );
+            who.setpos( dest );
+            return true;
+        },
+        [&who, is_u, seen]( const UncannyDodgeStatus fail ) -> bool {
```
Comment (chaosvolt) [inline src/character_functions.cpp:657]: ...I think this should eliminatethe rcs check correctly but please double-check my work on this function in particular beforehand should you opt to implement this.
```suggestion
    const auto visitor = cata::match{
        [&who, is_u, seen]( const UncannyDodgeStatus fail ) -> bool {
```

File: src/character_functions.cpp [hunk 6]
```diff
@@ -595,63 +599,140 @@ bool try_wield_contents( Character &who, item &container, item *internal_item, b
...
+                        add_msg( m_info, _( "%s tries to dodge but fails with a click!" ), who.disp_name() );
+                    }
+                    return false;
+                case UncannyDodgeStatus::NoSpace:
+                default:
+                    if( is_u ) {
+                        add_msg( _( "You try to dodge but there's no room!" ) );
+                    } else if( seen ) {
+                        add_msg( _( "%s tries to dodge but there's no room!" ), who.disp_name() );
+                    }
```
Comment (chaosvolt) [inline src/character_functions.cpp:680]:
```suggestion
                case UncannyDodgeStatus::Fail:
                default:
                    if( is_u ) {
                        add_msg( _( "You try to dodge but fail!" ) );
                    } else if( seen ) {
                        add_msg( _( "%s tries and fails to dodge!" ), who.disp_name() );
                    }
```

**Review by chaosvolt — DISMISSED** [review]
This should partially phase it out but hard to really sift through all these code changes to fully remove the RCS dodge.

**Comment by chaosvolt** [comment]
So, reverted the rocket system for now after discussion on the discord.
image

Move-dodge very rarely triggers, but when it does I already see a flaw in that it automatically smokes you before you finish leaving the tile:
image

**Comment by chaosvolt** [comment]
So, remaining musings on this:
1. The smoke thing still seems kinda hacky, plus right now if we want to keep it we need to log the old position and place the smoke there, instead the player smokes themselves every time they move-dodge. A better implementation would be to check `dodges_left` and subtract from it each time the player triggers a dodge. This would limit the maximum number of bullets per turn that can be dodged but not create a weird area limit.
2. The other obvious question then is, do we still want both skill-based dodge-in-place and move-dodge? Should skill-based dodges be unlimited and have failed rolls fall back on having move-dodges that are limited by number of dodges left that turn? Or should we go ahead and limit both types of dodge in this manner from the start? Or merge them so that you need both a successful skill roll and dodges left to dodge at all, but then always have the dodge swap your position?

**Comment by chaosvolt** [comment]
Okay, I'm testing some of those possible alternative ideas now and ran into a fun issue:
image

Move-dodge just outright makes you immune to the entire burst anyway, even if you moved forward and ended up still in the line of fire. So even if failed skill dodges were penalized by taking away from bonus dodges, move-dodge existing at all makes weapon fire basically useless against the user regardless of skill.

**Pull request closed without merging**